// @markuplint/ml-spec/lib/get-spec-by-tag-name.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSpecByTagName = void 0;
const get_attr_specs_1 = require("./get-attr-specs");
const cacheMap = new WeakMap();
function getSpecOM(spec) {
    const cache = cacheMap.get(spec);
    if (cache) {
        return cache;
    }
    const som = {};
    for (const el of spec.specs) {
        const attributes = (0, get_attr_specs_1.getAttrSpecs)(el.name, spec);
        som[el.name] = {
            experimental: !!el.experimental,
            obsolete: typeof el.obsolete === 'boolean' ? !!el.obsolete : el.obsolete ? el.obsolete.alt : false,
            deprecated: !!el.deprecated,
            nonStandard: !!el.nonStandard,
            categories: el.categories,
            permittedStructures: el.permittedStructures,
            attributes: Object.values(attributes || {}),
        };
    }
    cacheMap.set(spec, som);
    return som;
}
function getSpecByTagName(nameWithNS, specs) {
    const specOM = getSpecOM(specs);
    return specOM[nameWithNS] || null;
}
exports.getSpecByTagName = getSpecByTagName;
