// @pkgr/imagemin/lib/index.js
import { __awaiter } from "tslib";
import fs from 'node:fs';
const plugins = [
    [
        'gifsicle',
        {
            interlaced: true,
        },
    ],
    [
        'jpegtran',
        {
            progressive: true,
        },
    ],
    ['mozjpeg'],
    [
        'optipng',
        {
            optimizationLevel: 5,
        },
    ],
    [
        'pngquant',
        {
            // eslint-disable-next-line @typescript-eslint/no-magic-numbers
            quality: [0.6, 0.8],
        },
    ],
    ['upng'],
    [
        'svgo',
        {
            plugins: [
                {
                    name: 'preset-default',
                    params: {
                        overrides: {
                            removeViewBox: false,
                        },
                    },
                },
            ],
        },
    ],
    ['webp'],
].map(([name, opts]) => __awaiter(void 0, void 0, void 0, function* () {
    return (yield import(`imagemin-${name}`)).default(opts);
}));
export default (filename) => __awaiter(void 0, void 0, void 0, function* () {
    return fs.promises.writeFile(filename, yield plugins.reduce((acc, it) => __awaiter(void 0, void 0, void 0, function* () { return (yield it)(yield acc); }), fs.promises.readFile(filename)));
});
