// @markuplint/rules/lib/use-list/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ml_core_1 = require("@markuplint/ml-core");
const helpers_1 = require("../helpers");
exports.default = (0, ml_core_1.createRule)({
    defaultValue: [
        /**
         * @see https://en.wikipedia.org/wiki/Bullet_(typography)#In_Unicode
         */
        '\u2022',
        '\u2023',
        '\u2043',
        '\u204C',
        '\u204D',
        '\u2219',
        '\u25CB',
        '\u25CF',
        '\u25D8',
        '\u25E6',
        '\u2619',
        '\u2765',
        '\u2767',
        '\u29BE',
        '\u29BF',
        /**
         * In Japanese
         * @see https://ja.wikipedia.org/wiki/中黒#符号位置
         */
        '\u00B7',
        '\u0387',
        '\u2022',
        '\u2219',
        '\u22C5',
        '\u30FB',
        '\uFF65',
        /**
         * In Other Languages
         */
        /* REQUEST WANTED: https://github.com/markuplint/markuplint/issues/new */
        /**
         * From Markdown
         */
        '-',
        '*',
        '+', // plus signs
    ],
    defaultOptions: {
        spaceNeededBullets: [
            /**
             * From Markdown
             */
            '-',
            '*',
            '+', // plus signs
        ],
    },
    defaultServerity: 'warning',
    async verify({ document, report, t }) {
        await document.walkOn('Text', textNode => {
            const text = (0, helpers_1.decodeCharRef)(textNode.raw.trim());
            if (!text) {
                // emtpy
                return;
            }
            if (text.length === 1) {
                // character only
                return;
            }
            const bullets = textNode.rule.value;
            const spaceNeededBullets = textNode.rule.option.spaceNeededBullets || [];
            if (isMayListItem(text, bullets, spaceNeededBullets)) {
                report({
                    scope: textNode,
                    message: t('Use {0}', t('the {0}', t('{0*} element', 'li'))),
                });
            }
        });
    },
});
function isMayListItem(text, bullets, spaceNeededBullets) {
    const textArray = Array.from(text);
    const firstLetter = textArray[0];
    const isBullet = bullets.includes(firstLetter);
    const needSpace = spaceNeededBullets.includes(firstLetter);
    const continuous = firstLetter === textArray[1];
    if (continuous) {
        return false;
    }
    if (isBullet && needSpace) {
        const secondLetter = text[1];
        const isSpace = /^\s$/.test(secondLetter);
        return isSpace;
    }
    return isBullet;
}
