// @markuplint/ml-core/lib/ml-dom/node/document.js
"use strict";
var _MLDocument_filename, _MLDocument_tokenList;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MLDocument = void 0;
const tslib_1 = require("tslib");
const ml_config_1 = require("@markuplint/ml-config");
const ml_spec_1 = require("@markuplint/ml-spec");
const selector_1 = require("@markuplint/selector");
const debug_1 = require("../../debug");
const create_node_1 = require("../helper/create-node");
const debug_2 = require("../helper/debug");
const walkers_1 = require("../helper/walkers");
const node_list_1 = require("./node-list");
const parent_node_1 = require("./parent-node");
const rule_mapper_1 = require("./rule-mapper");
const unexpected_call_error_1 = tslib_1.__importDefault(require("./unexpected-call-error"));
const log = debug_1.log.extend('ml-dom');
const docLog = log.extend('document');
const ruleLog = docLog.extend('rule');
class MLDocument extends parent_node_1.MLParentNode {
    /**
     *
     * @param ast node list of markuplint AST
     * @param ruleset ruleset object
     */
    constructor(ast, ruleset, schemas, options) {
        var _a;
        // @ts-ignore
        super(ast, null);
        /**
         *
         */
        _MLDocument_filename.set(this, void 0);
        _MLDocument_tokenList.set(this, null);
        /**
         *
         */
        this.currentRule = null;
        this.isFragment = ast.isFragment;
        this.specs = (0, ml_spec_1.getSpec)(schemas);
        this.endTag = (_a = options === null || options === void 0 ? void 0 : options.endTag) !== null && _a !== void 0 ? _a : 'omittable';
        tslib_1.__classPrivateFieldSet(this, _MLDocument_filename, options === null || options === void 0 ? void 0 : options.filename, "f");
        // console.log(ast.nodeList.map((n, i) => `${i}: ${n.uuid} "${n.raw.trim()}"(${n.type})`));
        this.nodeList = Object.freeze(ast.nodeList
            .map(astNode => {
            if (astNode.type === 'endtag') {
                return;
            }
            return (0, create_node_1.createNode)(astNode, this);
        })
            .filter((n) => !!n));
        this._ruleMapping(ruleset);
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get activeElement() {
        throw new unexpected_call_error_1.default('Not supported "activeElement" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get alinkColor() {
        throw new unexpected_call_error_1.default('Not supported "alinkColor" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get all() {
        throw new unexpected_call_error_1.default('Not supported "all" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get anchors() {
        throw new unexpected_call_error_1.default('Not supported "anchors" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get applets() {
        throw new unexpected_call_error_1.default('Not supported "applets" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get bgColor() {
        throw new unexpected_call_error_1.default('Not supported "bgColor" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get body() {
        throw new unexpected_call_error_1.default('Not supported "body" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get characterSet() {
        throw new unexpected_call_error_1.default('Not supported "characterSet" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get charset() {
        throw new unexpected_call_error_1.default('Not supported "charset" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get compatMode() {
        throw new unexpected_call_error_1.default('Not supported "compatMode" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get contentType() {
        throw new unexpected_call_error_1.default('Not supported "contentType" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get cookie() {
        throw new unexpected_call_error_1.default('Not supported "cookie" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get currentScript() {
        throw new unexpected_call_error_1.default('Not supported "currentScript" property');
    }
    /**
     * Window object for calling the `getComputedStyle` and the `getPropertyValue` that
     * are needed by **Accessible Name and Description Computation**.
     * But it always returns the empty object.
     * (It may improve to possible to compute the name from the `style` attribute in the future.)
     *
     * @implements DOM API: `Document`
     */
    get defaultView() {
        return {
            getComputedStyle(_el) {
                return {
                    getPropertyValue(_propName) {
                        return {};
                    },
                };
            },
        };
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get designMode() {
        throw new unexpected_call_error_1.default('Not supported "designMode" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get dir() {
        throw new unexpected_call_error_1.default('Not supported "dir" property');
    }
    /**
     * @implements DOM API: `Document`
     */
    get doctype() {
        for (const node of this.nodeList) {
            if (node.is(node.DOCUMENT_TYPE_NODE)) {
                return node;
            }
        }
        return null;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get documentElement() {
        throw new unexpected_call_error_1.default('Not supported "documentElement" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get documentURI() {
        throw new unexpected_call_error_1.default('Not supported "documentURI" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get domain() {
        throw new unexpected_call_error_1.default('Not supported "domain" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get embeds() {
        throw new unexpected_call_error_1.default('Not supported "embeds" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get fgColor() {
        throw new unexpected_call_error_1.default('Not supported "fgColor" property');
    }
    /**
     * It could be used in rule, make sure it is immutable
     *
     * @implements `@markuplint/ml-core` API: `MLDOMDocument`
     */
    get filename() {
        return tslib_1.__classPrivateFieldGet(this, _MLDocument_filename, "f");
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get fonts() {
        throw new unexpected_call_error_1.default('Not supported "fonts" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get forms() {
        throw new unexpected_call_error_1.default('Not supported "forms" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get fullscreen() {
        throw new unexpected_call_error_1.default('Not supported "fullscreen" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get fullscreenElement() {
        throw new unexpected_call_error_1.default('Not supported "fullscreenElement" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get fullscreenEnabled() {
        throw new unexpected_call_error_1.default('Not supported "UfullscreenEnabledRL" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get head() {
        throw new unexpected_call_error_1.default('Not supported "head" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get hidden() {
        throw new unexpected_call_error_1.default('Not supported "hidden" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get images() {
        throw new unexpected_call_error_1.default('Not supported "images" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get implementation() {
        throw new unexpected_call_error_1.default('Not supported "implementation" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get inputEncoding() {
        throw new unexpected_call_error_1.default('Not supported "inputEncoding" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get lastModified() {
        throw new unexpected_call_error_1.default('Not supported "lastModified" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get linkColor() {
        throw new unexpected_call_error_1.default('Not supported "linkColor" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get links() {
        throw new unexpected_call_error_1.default('Not supported "links" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get location() {
        throw new unexpected_call_error_1.default('Not supported "location" property');
    }
    /**
     * Returns a string appropriate for the type of node as `Document`
     *
     * @see https://dom.spec.whatwg.org/#ref-for-document%E2%91%A8
     */
    get nodeName() {
        return '#document';
    }
    /**
     * Returns a number appropriate for the type of `Document`
     */
    get nodeType() {
        return this.DOCUMENT_NODE;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onabort() {
        throw new unexpected_call_error_1.default('Not supported "onabort" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onanimationcancel() {
        throw new unexpected_call_error_1.default('Not supported "onanimationcancel" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onanimationend() {
        throw new unexpected_call_error_1.default('Not supported "onanimationend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onanimationiteration() {
        throw new unexpected_call_error_1.default('Not supported "onanimationiteration" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onanimationstart() {
        throw new unexpected_call_error_1.default('Not supported "onanimationstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onauxclick() {
        throw new unexpected_call_error_1.default('Not supported "onauxclick" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onblur() {
        throw new unexpected_call_error_1.default('Not supported "onblur" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get oncanplay() {
        throw new unexpected_call_error_1.default('Not supported "oncanplay" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get oncanplaythrough() {
        throw new unexpected_call_error_1.default('Not supported "oncanplaythrough" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onchange() {
        throw new unexpected_call_error_1.default('Not supported "onchange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onclick() {
        throw new unexpected_call_error_1.default('Not supported "onclick" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onclose() {
        throw new unexpected_call_error_1.default('Not supported "onclose" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get oncontextmenu() {
        throw new unexpected_call_error_1.default('Not supported "oncontextmenu" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get oncopy() {
        throw new unexpected_call_error_1.default('Not supported "oncopy" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get oncuechange() {
        throw new unexpected_call_error_1.default('Not supported "oncuechange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get oncut() {
        throw new unexpected_call_error_1.default('Not supported "oncut" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ondblclick() {
        throw new unexpected_call_error_1.default('Not supported "ondblclick" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ondrag() {
        throw new unexpected_call_error_1.default('Not supported "ondrag" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ondragend() {
        throw new unexpected_call_error_1.default('Not supported "ondragend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ondragenter() {
        throw new unexpected_call_error_1.default('Not supported "ondragenter" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ondragleave() {
        throw new unexpected_call_error_1.default('Not supported "ondragleave" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ondragover() {
        throw new unexpected_call_error_1.default('Not supported "ondragover" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ondragstart() {
        throw new unexpected_call_error_1.default('Not supported "ondragstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ondrop() {
        throw new unexpected_call_error_1.default('Not supported "ondrop" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ondurationchange() {
        throw new unexpected_call_error_1.default('Not supported "ondurationchange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onemptied() {
        throw new unexpected_call_error_1.default('Not supported "onemptied" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onended() {
        throw new unexpected_call_error_1.default('Not supported "onended" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onerror() {
        throw new unexpected_call_error_1.default('Not supported "onerror" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onfocus() {
        throw new unexpected_call_error_1.default('Not supported "onfocus" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onformdata() {
        throw new unexpected_call_error_1.default('Not supported "onformdata" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onfullscreenchange() {
        throw new unexpected_call_error_1.default('Not supported "onfullscreenchange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onfullscreenerror() {
        throw new unexpected_call_error_1.default('Not supported "onfullscreenerror" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ongotpointercapture() {
        throw new unexpected_call_error_1.default('Not supported "ongotpointercapture" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get oninput() {
        throw new unexpected_call_error_1.default('Not supported "oninput" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get oninvalid() {
        throw new unexpected_call_error_1.default('Not supported "oninvalid" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onkeydown() {
        throw new unexpected_call_error_1.default('Not supported "onkeydown" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onkeypress() {
        throw new unexpected_call_error_1.default('Not supported "onkeypress" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onkeyup() {
        throw new unexpected_call_error_1.default('Not supported "onkeyup" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onload() {
        throw new unexpected_call_error_1.default('Not supported "onload" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onloadeddata() {
        throw new unexpected_call_error_1.default('Not supported "onloadeddata" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onloadedmetadata() {
        throw new unexpected_call_error_1.default('Not supported "onloadedmetadata" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onloadstart() {
        throw new unexpected_call_error_1.default('Not supported "onloadstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onlostpointercapture() {
        throw new unexpected_call_error_1.default('Not supported "onlostpointercapture" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onmousedown() {
        throw new unexpected_call_error_1.default('Not supported "onmousedown" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onmouseenter() {
        throw new unexpected_call_error_1.default('Not supported "onmouseenter" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onmouseleave() {
        throw new unexpected_call_error_1.default('Not supported "onmouseleave" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onmousemove() {
        throw new unexpected_call_error_1.default('Not supported "onmousemove" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onmouseout() {
        throw new unexpected_call_error_1.default('Not supported "onmouseout" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onmouseover() {
        throw new unexpected_call_error_1.default('Not supported "onmouseover" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onmouseup() {
        throw new unexpected_call_error_1.default('Not supported "onmouseup" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpaste() {
        throw new unexpected_call_error_1.default('Not supported "onpaste" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpause() {
        throw new unexpected_call_error_1.default('Not supported "onpause" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onplay() {
        throw new unexpected_call_error_1.default('Not supported "onplay" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onplaying() {
        throw new unexpected_call_error_1.default('Not supported "onplaying" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpointercancel() {
        throw new unexpected_call_error_1.default('Not supported "onpointercancel" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpointerdown() {
        throw new unexpected_call_error_1.default('Not supported "onpointerdown" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpointerenter() {
        throw new unexpected_call_error_1.default('Not supported "onpointerenter" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpointerleave() {
        throw new unexpected_call_error_1.default('Not supported "onpointerleave" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpointerlockchange() {
        throw new unexpected_call_error_1.default('Not supported "onpointerlockchange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpointerlockerror() {
        throw new unexpected_call_error_1.default('Not supported "onpointerlockerror" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpointermove() {
        throw new unexpected_call_error_1.default('Not supported "onpointermove" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpointerout() {
        throw new unexpected_call_error_1.default('Not supported "onpointerout" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpointerover() {
        throw new unexpected_call_error_1.default('Not supported "onpointerover" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onpointerup() {
        throw new unexpected_call_error_1.default('Not supported "onpointerup" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onprogress() {
        throw new unexpected_call_error_1.default('Not supported "onprogress" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onratechange() {
        throw new unexpected_call_error_1.default('Not supported "onratechange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onreadystatechange() {
        throw new unexpected_call_error_1.default('Not supported "onreadystatechange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onreset() {
        throw new unexpected_call_error_1.default('Not supported "onreset" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onresize() {
        throw new unexpected_call_error_1.default('Not supported "onresize" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onscroll() {
        throw new unexpected_call_error_1.default('Not supported "onscroll" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onsecuritypolicyviolation() {
        throw new unexpected_call_error_1.default('Not supported "onsecuritypolicyviolation" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onseeked() {
        throw new unexpected_call_error_1.default('Not supported "onseeked" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onseeking() {
        throw new unexpected_call_error_1.default('Not supported "onseeking" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onselect() {
        throw new unexpected_call_error_1.default('Not supported "onselect" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onselectionchange() {
        throw new unexpected_call_error_1.default('Not supported "onselectionchange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onselectstart() {
        throw new unexpected_call_error_1.default('Not supported "onselectstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onslotchange() {
        throw new unexpected_call_error_1.default('Not supported "onslotchange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onstalled() {
        throw new unexpected_call_error_1.default('Not supported "onstalled" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onsubmit() {
        throw new unexpected_call_error_1.default('Not supported "onsubmit" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onsuspend() {
        throw new unexpected_call_error_1.default('Not supported "onsuspend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ontimeupdate() {
        throw new unexpected_call_error_1.default('Not supported "ontimeupdate" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ontoggle() {
        throw new unexpected_call_error_1.default('Not supported "ontoggle" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ontransitioncancel() {
        throw new unexpected_call_error_1.default('Not supported "ontransitioncancel" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ontransitionend() {
        throw new unexpected_call_error_1.default('Not supported "ontransitionend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ontransitionrun() {
        throw new unexpected_call_error_1.default('Not supported "ontransitionrun" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get ontransitionstart() {
        throw new unexpected_call_error_1.default('Not supported "ontransitionstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onvisibilitychange() {
        throw new unexpected_call_error_1.default('Not supported "onvisibilitychange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onvolumechange() {
        throw new unexpected_call_error_1.default('Not supported "onvolumechange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onwaiting() {
        throw new unexpected_call_error_1.default('Not supported "onwaiting" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onwebkitanimationend() {
        throw new unexpected_call_error_1.default('Not supported "onwebkitanimationend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onwebkitanimationiteration() {
        throw new unexpected_call_error_1.default('Not supported "onwebkitanimationiteration" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onwebkitanimationstart() {
        throw new unexpected_call_error_1.default('Not supported "onwebkitanimationstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onwebkittransitionend() {
        throw new unexpected_call_error_1.default('Not supported "onwebkittransitionend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get onwheel() {
        throw new unexpected_call_error_1.default('Not supported "onwheel" property');
    }
    get ownerDocument() {
        return null;
    }
    get ownerMLDocument() {
        return this;
    }
    /**
     * @implements DOM API: `Document`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-node-parentnode%E2%91%A0
     */
    get parentNode() {
        return null;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get pictureInPictureElement() {
        throw new unexpected_call_error_1.default('Not supported "pictureInPictureElement" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get pictureInPictureEnabled() {
        throw new unexpected_call_error_1.default('Not supported "pictureInPictureEnabled" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get plugins() {
        throw new unexpected_call_error_1.default('Not supported "plugins" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get pointerLockElement() {
        throw new unexpected_call_error_1.default('Not supported "pointerLockElement" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get readyState() {
        throw new unexpected_call_error_1.default('Not supported "readyState" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get referrer() {
        throw new unexpected_call_error_1.default('Not supported "referrer" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get rootElement() {
        throw new unexpected_call_error_1.default('Not supported "rootElement" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get scripts() {
        throw new unexpected_call_error_1.default('Not supported "scripts" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get scrollingElement() {
        throw new unexpected_call_error_1.default('Not supported "scrollingElement" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get styleSheets() {
        throw new unexpected_call_error_1.default('Not supported "styleSheets" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get timeline() {
        throw new unexpected_call_error_1.default('Not supported "timeline" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get title() {
        throw new unexpected_call_error_1.default('Not supported "title" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get URL() {
        throw new unexpected_call_error_1.default('Not supported "URL" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    get visibilityState() {
        throw new unexpected_call_error_1.default('Not supported "visibilityState" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    get vlinkColor() {
        throw new unexpected_call_error_1.default('Not supported "vlinkColor" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    adoptNode(node) {
        throw new unexpected_call_error_1.default('Not supported "adoptNode" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    captureEvents() {
        throw new unexpected_call_error_1.default('Not supported "captureEvents" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    caretRangeFromPoint(x, y) {
        throw new unexpected_call_error_1.default('Not supported "caretRangeFromPoint" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    clear() {
        throw new unexpected_call_error_1.default('Not supported "clear" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    close() {
        throw new unexpected_call_error_1.default('Not supported "close" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createAttribute(localName) {
        throw new unexpected_call_error_1.default('Not supported "createAttribute" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createAttributeNS(namespace, qualifiedName) {
        throw new unexpected_call_error_1.default('Not supported "createAttributeNS" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createCDATASection(data) {
        throw new unexpected_call_error_1.default('Not supported "createCDATASection" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createComment(data) {
        throw new unexpected_call_error_1.default('Not supported "createComment" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createDocumentFragment() {
        throw new unexpected_call_error_1.default('Not supported "createDocumentFragment" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createElement(tagName, options) {
        throw new unexpected_call_error_1.default('Not supported "createElement" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createElementNS(namespace, qualifiedName, options) {
        throw new unexpected_call_error_1.default('Not supported "createElementNS" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createEvent(eventInterface) {
        throw new unexpected_call_error_1.default('Not supported "createEvent" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createExpression(expression, resolver) {
        throw new unexpected_call_error_1.default('Not supported "createExpression" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createNodeIterator(root, whatToShow, filter) {
        throw new unexpected_call_error_1.default('Not supported "createNodeIterator" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createNSResolver(nodeResolver) {
        throw new unexpected_call_error_1.default('Not supported "createNSResolver" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createProcessingInstruction(target, data) {
        throw new unexpected_call_error_1.default('Not supported "createProcessingInstruction" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createRange() {
        throw new unexpected_call_error_1.default('Not supported "createRange" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createTextNode(data) {
        throw new unexpected_call_error_1.default('Not supported "createTextNode" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    createTreeWalker(root, whatToShow, filter) {
        throw new unexpected_call_error_1.default('Not supported "createTreeWalker" method');
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDocument`
     */
    debugMap() {
        return (0, debug_2.nodeListToDebugMaps)(this.nodeList, true);
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    elementFromPoint(x, y) {
        throw new unexpected_call_error_1.default('Not supported "elementFromPoint" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    elementsFromPoint(x, y) {
        throw new unexpected_call_error_1.default('Not supported "elementsFromPoint" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    evaluate(expression, contextNode, resolver, type, result) {
        throw new unexpected_call_error_1.default('Not supported "evaluate" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    execCommand(commandId, showUI, value) {
        throw new unexpected_call_error_1.default('Not supported "execCommand" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    exitFullscreen() {
        throw new unexpected_call_error_1.default('Not supported "exitFullscreen" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    exitPictureInPicture() {
        throw new unexpected_call_error_1.default('Not supported "exitPictureInPicture" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    exitPointerLock() {
        throw new unexpected_call_error_1.default('Not supported "exitPointerLock" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    getAnimations() {
        throw new unexpected_call_error_1.default('Not supported "getAnimations" method');
    }
    /**
     * @implements DOM API: `Document`
     */
    getElementById(elementId) {
        // TODO:
        return this.querySelector(`#${elementId}`);
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    getElementsByClassName(classNames) {
        return (0, node_list_1.nodeListToHTMLCollection)(this.querySelectorAll(classNames));
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    getElementsByName(elementName) {
        throw new unexpected_call_error_1.default('Not supported "getElementsByName" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    getElementsByTagName(qualifiedName) {
        return (0, node_list_1.nodeListToHTMLCollection)(this.querySelectorAll(qualifiedName));
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    getElementsByTagNameNS(namespace, localName) {
        throw new unexpected_call_error_1.default('Not supported "getElementsByTagNameNS" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    getSelection() {
        throw new unexpected_call_error_1.default('Not supported "getSelection" method');
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDocument`
     */
    getTokenList() {
        if (tslib_1.__classPrivateFieldGet(this, _MLDocument_tokenList, "f")) {
            return tslib_1.__classPrivateFieldGet(this, _MLDocument_tokenList, "f");
        }
        const tokens = [];
        for (const node of this.nodeList) {
            tokens.push(node);
            if (node.is(node.ELEMENT_NODE) && node.closeTag) {
                tokens.push(node.closeTag);
            }
        }
        tokens.sort((a, b) => a.startOffset - b.startOffset);
        tslib_1.__classPrivateFieldSet(this, _MLDocument_tokenList, Object.freeze(tokens), "f");
        return tslib_1.__classPrivateFieldGet(this, _MLDocument_tokenList, "f");
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    hasFocus() {
        throw new unexpected_call_error_1.default('Not supported "hasFocus" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    hasStorageAccess() {
        throw new unexpected_call_error_1.default('Not supported "hasStorageAccess" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    importNode(node, deep) {
        throw new unexpected_call_error_1.default('Not supported "importNode" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    open(url, name, features) {
        throw new unexpected_call_error_1.default('Not supported "open" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    queryCommandEnabled(commandId) {
        throw new unexpected_call_error_1.default('Not supported "queryCommandEnabled" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    queryCommandIndeterm(commandId) {
        throw new unexpected_call_error_1.default('Not supported "queryCommandIndeterm" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    queryCommandState(commandId) {
        throw new unexpected_call_error_1.default('Not supported "queryCommandState" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    queryCommandSupported(commandId) {
        throw new unexpected_call_error_1.default('Not supported "queryCommandSupported" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    queryCommandValue(commandId) {
        throw new unexpected_call_error_1.default('Not supported "queryCommandValue" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Document`
     */
    releaseEvents() {
        throw new unexpected_call_error_1.default('Not supported "releaseEvents" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    requestStorageAccess() {
        throw new unexpected_call_error_1.default('Not supported "requestStorageAccess" method');
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDocument`
     */
    setRule(rule) {
        this.currentRule = rule;
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDocument`
     */
    toString() {
        const html = [];
        for (const node of this.getTokenList()) {
            html.push(node.toString());
        }
        return html.join('');
    }
    walkOn(type, walker, skipWhenRuleIsDisabled = true) {
        return (0, walkers_1.sequentialWalker)(this.nodeList, node => {
            if (skipWhenRuleIsDisabled && node.rule.disabled) {
                return;
            }
            if (type === 'Element' && node.is(node.ELEMENT_NODE)) {
                return walker(node);
            }
            if (type === 'Text' && node.is(node.TEXT_NODE)) {
                return walker(node);
            }
            if (type === 'Comment' && node.is(node.COMMENT_NODE)) {
                return walker(node);
            }
            if (type === 'Attr' && node.is(node.ELEMENT_NODE)) {
                return (0, walkers_1.sequentialWalker)(Array.from(node.attributes), attr => walker(attr));
            }
            if (type === 'ElementCloseTag' && node.is(node.ELEMENT_NODE) && node.closeTag) {
                return walker(node.closeTag);
            }
        });
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    write(...text) {
        throw new unexpected_call_error_1.default('Not supported "write" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Document`
     */
    writeln(...text) {
        throw new unexpected_call_error_1.default('Not supported "writeln" method');
    }
    _ruleMapping(ruleset) {
        if (docLog.enabled) {
            docLog('Rule Mapping: %O', Object.keys(ruleset.rules));
        }
        const ruleMapper = new rule_mapper_1.RuleMapper(this);
        // global rules to #document
        Object.keys(ruleset.rules).forEach(ruleName => {
            const rule = ruleset.rules[ruleName];
            ruleMapper.set(this, ruleName, {
                from: 'rules',
                specificity: [0, 0, 0],
                rule,
            });
        });
        // add rules to node
        for (const node of this.nodeList) {
            if (docLog.enabled) {
                docLog('Add rules to node <%s>', node.nodeName);
            }
            // global rules to each element
            Object.keys(ruleset.rules).forEach(ruleName => {
                const rule = ruleset.rules[ruleName];
                ruleMapper.set(node, ruleName, {
                    from: 'rules',
                    specificity: [0, 0, 0],
                    rule,
                });
            });
            if (!node.is(node.ELEMENT_NODE) && !node.is(node.TEXT_NODE)) {
                continue;
            }
            const selectorTarget = node.is(node.ELEMENT_NODE) ? node : null;
            // node specs and special rules for node by selector
            ruleset.nodeRules.forEach(nodeRule => {
                if (!nodeRule.rules) {
                    return;
                }
                if (!selectorTarget) {
                    return;
                }
                const selector = nodeRule.selector || nodeRule.regexSelector;
                const matches = (0, selector_1.matchSelector)(selectorTarget, selector);
                if (!matches.matched) {
                    return;
                }
                if (docLog.enabled) {
                    docLog('Matched nodeRule: <%s>(%s)', node.nodeName, matches.selector || '*');
                }
                const ruleList = Object.keys(nodeRule.rules);
                for (const ruleName of ruleList) {
                    const rule = nodeRule.rules[ruleName];
                    const convertedRule = (0, ml_config_1.exchangeValueOnRule)(rule, matches.data || {});
                    if (convertedRule === undefined) {
                        continue;
                    }
                    const globalRule = ruleset.rules[ruleName];
                    const mergedRule = globalRule ? (0, ml_config_1.mergeRule)(globalRule, convertedRule) : convertedRule;
                    ruleLog(' nodeRule (%s): %O', ruleName, mergedRule);
                    ruleMapper.set(node, ruleName, {
                        from: 'nodeRules',
                        specificity: matches.specificity,
                        rule: mergedRule,
                    });
                }
            });
            // overwrite rule to child node
            if (selectorTarget && ruleset.childNodeRules.length) {
                const descendants = [];
                const children = Array.from(selectorTarget.childNodes);
                (0, walkers_1.syncWalk)(children, childNode => {
                    descendants.push(childNode);
                });
                ruleset.childNodeRules.forEach((nodeRule, i) => {
                    if (!nodeRule.rules) {
                        return;
                    }
                    const nodeRuleRules = nodeRule.rules;
                    const selector = nodeRule.selector || nodeRule.regexSelector;
                    if (!selector) {
                        return;
                    }
                    const matches = (0, selector_1.matchSelector)(selectorTarget, selector);
                    if (!matches.matched) {
                        return;
                    }
                    if (docLog.enabled) {
                        docLog('Matched childNodeRule: <%s>(%s), inheritance: %o', selectorTarget.nodeName, matches.selector || '*', !!nodeRule.inheritance);
                    }
                    const targetDescendants = nodeRule.inheritance ? descendants : children;
                    Object.keys(nodeRuleRules).forEach(ruleName => {
                        const rule = nodeRuleRules[ruleName];
                        const convertedRule = (0, ml_config_1.exchangeValueOnRule)(rule, matches.data || {});
                        if (convertedRule === undefined) {
                            return;
                        }
                        const globalRule = ruleset.rules[ruleName];
                        const mergedRule = globalRule ? (0, ml_config_1.mergeRule)(globalRule, convertedRule) : convertedRule;
                        ruleLog(' childNodeRule (%s): %O', ruleName, mergedRule);
                        targetDescendants.forEach(descendant => {
                            ruleMapper.set(descendant, ruleName, {
                                from: 'childNodeRules',
                                specificity: matches.specificity,
                                rule: mergedRule,
                            });
                        });
                    });
                });
            }
        }
        ruleMapper.apply();
    }
}
exports.MLDocument = MLDocument;
_MLDocument_filename = new WeakMap(), _MLDocument_tokenList = new WeakMap();
