// @markuplint/ml-core/lib/ml-dom/node/node-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeListToHTMLCollection = exports.toHTMLCollection = exports.toNodeList = void 0;
class MLNodeList extends Array {
    forEach(callbackfn, thisArg) {
        return super.forEach.bind(this)((v, k) => callbackfn(v, k, thisArg !== null && thisArg !== void 0 ? thisArg : this));
    }
    item(index) {
        return this[index];
    }
}
function toNodeList(nodes) {
    const nodeList = new MLNodeList(...nodes);
    return nodeList;
}
exports.toNodeList = toNodeList;
class MLHTMLCollection extends Array {
    item(index) {
        return this[index];
    }
    namedItem(name) {
        return this.find(el => el.getAttribute('name') === name) || null;
    }
}
function toHTMLCollection(nodes) {
    const collection = new MLHTMLCollection(...nodes);
    return collection;
}
exports.toHTMLCollection = toHTMLCollection;
function nodeListToHTMLCollection(nodeList) {
    const collection = new MLHTMLCollection();
    nodeList.forEach(node => {
        if (node.is(node.ELEMENT_NODE)) {
            collection.push(node);
        }
    });
    return collection;
}
exports.nodeListToHTMLCollection = nodeListToHTMLCollection;
