// @markuplint/ml-core/lib/ml-dom/node/node-store.js
"use strict";
var _NodeStore_store;
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeStore = void 0;
const tslib_1 = require("tslib");
const parser_utils_1 = require("@markuplint/parser-utils");
class NodeStore {
    constructor() {
        _NodeStore_store.set(this, new Map());
    }
    getNode(astNode) {
        // console.log(`Get: ${astNode.uuid} -> ${astNode.raw.trim()}(${astNode.type})`);
        const node = tslib_1.__classPrivateFieldGet(this, _NodeStore_store, "f").get(astNode.uuid);
        if (!node) {
            throw new parser_utils_1.ParserError('Broke mapping nodes.', {
                line: astNode.startLine,
                col: astNode.startCol,
                raw: astNode.raw,
                nodeName: astNode.nodeName,
            });
        }
        return node;
    }
    setNode(astNode, node) {
        tslib_1.__classPrivateFieldGet(this, _NodeStore_store, "f").set(astNode.uuid, node);
    }
}
_NodeStore_store = new WeakMap();
/**
 * `NodeStore` Singleton
 */
exports.nodeStore = new NodeStore();
