// @markuplint/ml-core/lib/ml-dom/token/token.js
"use strict";
var _MLToken_endCol, _MLToken_endLine, _MLToken_endOffset, _MLToken_fixed, _MLToken_raw, _MLToken_startCol, _MLToken_startLine, _MLToken_startOffset;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MLToken = void 0;
const tslib_1 = require("tslib");
class MLToken {
    constructor(astToken) {
        _MLToken_endCol.set(this, void 0);
        _MLToken_endLine.set(this, void 0);
        _MLToken_endOffset.set(this, void 0);
        _MLToken_fixed.set(this, void 0);
        _MLToken_raw.set(this, void 0);
        _MLToken_startCol.set(this, void 0);
        _MLToken_startLine.set(this, void 0);
        _MLToken_startOffset.set(this, void 0);
        this._astToken = astToken;
        tslib_1.__classPrivateFieldSet(this, _MLToken_raw, astToken.raw, "f");
        tslib_1.__classPrivateFieldSet(this, _MLToken_fixed, astToken.raw, "f");
        this.uuid = astToken.uuid;
        tslib_1.__classPrivateFieldSet(this, _MLToken_startLine, astToken.startLine, "f");
        tslib_1.__classPrivateFieldSet(this, _MLToken_endLine, astToken.endLine, "f");
        tslib_1.__classPrivateFieldSet(this, _MLToken_startCol, astToken.startCol, "f");
        tslib_1.__classPrivateFieldSet(this, _MLToken_endCol, astToken.endCol, "f");
        tslib_1.__classPrivateFieldSet(this, _MLToken_startOffset, astToken.startOffset, "f");
        tslib_1.__classPrivateFieldSet(this, _MLToken_endOffset, astToken.endOffset, "f");
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDOMToken`
     */
    get endCol() {
        return tslib_1.__classPrivateFieldGet(this, _MLToken_endCol, "f");
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDOMToken`
     */
    get endLine() {
        return tslib_1.__classPrivateFieldGet(this, _MLToken_endLine, "f");
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDOMToken`
     */
    get endOffset() {
        return tslib_1.__classPrivateFieldGet(this, _MLToken_endOffset, "f");
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDOMToken`
     */
    get originRaw() {
        return tslib_1.__classPrivateFieldGet(this, _MLToken_raw, "f");
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDOMToken`
     */
    get raw() {
        return tslib_1.__classPrivateFieldGet(this, _MLToken_fixed, "f");
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDOMToken`
     */
    get startCol() {
        return tslib_1.__classPrivateFieldGet(this, _MLToken_startCol, "f");
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDOMToken`
     */
    get startLine() {
        return tslib_1.__classPrivateFieldGet(this, _MLToken_startLine, "f");
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDOMToken`
     */
    get startOffset() {
        return tslib_1.__classPrivateFieldGet(this, _MLToken_startOffset, "f");
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDOMToken`
     */
    fix(raw) {
        tslib_1.__classPrivateFieldSet(this, _MLToken_fixed, raw, "f");
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLDOMToken`
     */
    toString() {
        return this.raw;
    }
}
exports.MLToken = MLToken;
_MLToken_endCol = new WeakMap(), _MLToken_endLine = new WeakMap(), _MLToken_endOffset = new WeakMap(), _MLToken_fixed = new WeakMap(), _MLToken_raw = new WeakMap(), _MLToken_startCol = new WeakMap(), _MLToken_startLine = new WeakMap(), _MLToken_startOffset = new WeakMap();
