// @markuplint/rules/lib/no-default-value/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ml_core_1 = require("@markuplint/ml-core");
const helpers_1 = require("../helpers");
exports.default = (0, ml_core_1.createRule)({
    defaultServerity: 'warning',
    async verify({ document, report, t }) {
        await document.walkOn('Element', el => {
            const attrSpec = (0, ml_core_1.getAttrSpecs)(el.nameWithNS, document.specs);
            if (!attrSpec) {
                return;
            }
            for (const attr of el.attributes) {
                if (attr.attrType !== 'html-attr') {
                    continue;
                }
                const spec = attrSpec.find(s => s.name === attr.getName().potential);
                if (!spec) {
                    continue;
                }
                if (!spec.defaultValue) {
                    continue;
                }
                const normalizedValue = (0, helpers_1.toNormalizedValue)(attr.getValue().potential, spec);
                const defaultValue = (0, helpers_1.toNormalizedValue)(spec.defaultValue, spec);
                if (defaultValue === normalizedValue) {
                    report({
                        scope: el,
                        line: attr.valueNode.startLine,
                        col: attr.valueNode.startCol,
                        raw: attr.valueNode.raw,
                        message: t('It is {0}', t('the {0}', 'default value')),
                    });
                }
            }
        });
    },
});
