// @markuplint/ml-core/lib/ml-rule/ml-rule-context.js
"use strict";
var _MLRuleContext_reports;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MLRuleContext = void 0;
const tslib_1 = require("tslib");
const i18n_1 = require("@markuplint/i18n");
class MLRuleContext {
    constructor(document, locale, rule) {
        _MLRuleContext_reports.set(this, []);
        this.document = document;
        this.globalRule = rule;
        this.translate = (0, i18n_1.translator)(locale);
        this.locale = locale.locale;
    }
    get reports() {
        return tslib_1.__classPrivateFieldGet(this, _MLRuleContext_reports, "f").map(report => ({
            ...report,
            message: finish(report.message),
        }));
    }
    report(report) {
        tslib_1.__classPrivateFieldGet(this, _MLRuleContext_reports, "f").push(report);
    }
    provide() {
        return {
            document: this.document,
            translate: this.translate,
            t: this.translate,
            globalRule: this.globalRule,
            reports: this.reports,
            report: this.report.bind(this),
        };
    }
}
exports.MLRuleContext = MLRuleContext;
_MLRuleContext_reports = new WeakMap();
function finish(message, locale = 'en') {
    switch (locale) {
        case 'en': {
            return message.replace(/^[a-z]/, $0 => $0.toUpperCase());
        }
    }
    return message;
}
