// @markuplint/parser-utils/lib/get-location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sliceFragment = exports.getEndCol = exports.getEndLine = exports.getCol = exports.getLine = void 0;
function getLine(html, startOffset) {
    return html.slice(0, startOffset).split(/\n/g).length;
}
exports.getLine = getLine;
function getCol(html, startOffset) {
    const lines = html.slice(0, startOffset).split(/\n/g);
    return lines[lines.length - 1].length + 1;
}
exports.getCol = getCol;
function getEndLine(html, line) {
    return html.split(/\r?\n/).length - 1 + line;
}
exports.getEndLine = getEndLine;
function getEndCol(html, col) {
    const lines = html.split(/\r?\n/);
    const lineCount = lines.length;
    const lastLine = lines.pop();
    return lineCount > 1 ? lastLine.length + 1 : col + html.length;
}
exports.getEndCol = getEndCol;
function sliceFragment(rawHtml, start, end) {
    const raw = rawHtml.slice(start, end);
    return {
        startOffset: start,
        endOffset: end,
        startLine: getLine(rawHtml, start),
        endLine: getLine(rawHtml, end),
        startCol: getCol(rawHtml, start),
        endCol: getCol(rawHtml, end),
        raw,
    };
}
exports.sliceFragment = sliceFragment;
