// @markuplint/types/lib/primitive/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.range = exports.splitUnit = exports.isQuantity = exports.isNonZeroUint = exports.isFloat = exports.isUint = exports.isInt = void 0;
var is_int_1 = require("./is-int");
Object.defineProperty(exports, "isInt", { enumerable: true, get: function () { return is_int_1.isInt; } });
var is_uint_1 = require("./is-uint");
Object.defineProperty(exports, "isUint", { enumerable: true, get: function () { return is_uint_1.isUint; } });
var is_float_1 = require("./is-float");
Object.defineProperty(exports, "isFloat", { enumerable: true, get: function () { return is_float_1.isFloat; } });
var is_non_zero_uint_1 = require("./is-non-zero-uint");
Object.defineProperty(exports, "isNonZeroUint", { enumerable: true, get: function () { return is_non_zero_uint_1.isNonZeroUint; } });
var is_quantity_1 = require("./is-quantity");
Object.defineProperty(exports, "isQuantity", { enumerable: true, get: function () { return is_quantity_1.isQuantity; } });
var split_unit_1 = require("./split-unit");
Object.defineProperty(exports, "splitUnit", { enumerable: true, get: function () { return split_unit_1.splitUnit; } });
var range_1 = require("./range");
Object.defineProperty(exports, "range", { enumerable: true, get: function () { return range_1.range; } });
