// @markuplint/ml-core/lib/ml-dom/tokens/doctype.js
"use strict";
var _MLDOMDoctype_name, _MLDOMDoctype_publicId, _MLDOMDoctype_systemId;
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const node_1 = tslib_1.__importDefault(require("./node"));
class MLDOMDoctype extends node_1.default {
    constructor(astNode, document) {
        super(astNode, document);
        this.nodeType = 10;
        this.type = 'Doctype';
        _MLDOMDoctype_name.set(this, void 0);
        _MLDOMDoctype_publicId.set(this, void 0);
        _MLDOMDoctype_systemId.set(this, void 0);
        tslib_1.__classPrivateFieldSet(this, _MLDOMDoctype_name, astNode.name, "f");
        tslib_1.__classPrivateFieldSet(this, _MLDOMDoctype_publicId, astNode.publicId, "f");
        tslib_1.__classPrivateFieldSet(this, _MLDOMDoctype_systemId, astNode.systemId, "f");
    }
    get name() {
        return tslib_1.__classPrivateFieldGet(this, _MLDOMDoctype_name, "f");
    }
    get publicId() {
        return tslib_1.__classPrivateFieldGet(this, _MLDOMDoctype_publicId, "f");
    }
    get systemId() {
        return tslib_1.__classPrivateFieldGet(this, _MLDOMDoctype_systemId, "f");
    }
    get textContent() {
        return null;
    }
}
exports.default = MLDOMDoctype;
_MLDOMDoctype_name = new WeakMap(), _MLDOMDoctype_publicId = new WeakMap(), _MLDOMDoctype_systemId = new WeakMap();
