// @markuplint/ml-core/lib/ml-dom/tokens/preprocessor-specific-attribute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const token_1 = tslib_1.__importDefault(require("./token"));
class MLDOMPreprocessorSpecificAttribute extends token_1.default {
    constructor(astToken) {
        super(astToken);
        this.attrType = 'ps-attr';
        this.potentialName = astToken.potentialName;
        this.potentialValue = astToken.potentialValue;
        this.valueType = astToken.valueType;
        this.isDuplicatable = astToken.isDuplicatable;
    }
    getName() {
        return {
            line: this.startLine,
            col: this.startCol,
            potential: this.potentialName.toLowerCase(),
            raw: this.raw,
        };
    }
    getValue() {
        return {
            line: this.startLine,
            col: this.startCol,
            potential: this.potentialValue,
            raw: this.raw,
        };
    }
    toString() {
        return this.raw;
    }
    toNormalizeString() {
        return this.raw;
    }
}
exports.default = MLDOMPreprocessorSpecificAttribute;
