// @markuplint/ml-core/lib/ml-dom/tokens/token.js
"use strict";
var _MLDOMToken_startLine, _MLDOMToken_endLine, _MLDOMToken_startCol, _MLDOMToken_endCol, _MLDOMToken_startOffset, _MLDOMToken_endOffset, _MLDOMToken_raw, _MLDOMToken_fixed;
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
class MLDOMToken {
    constructor(astToken) {
        _MLDOMToken_startLine.set(this, void 0);
        _MLDOMToken_endLine.set(this, void 0);
        _MLDOMToken_startCol.set(this, void 0);
        _MLDOMToken_endCol.set(this, void 0);
        _MLDOMToken_startOffset.set(this, void 0);
        _MLDOMToken_endOffset.set(this, void 0);
        _MLDOMToken_raw.set(this, void 0);
        _MLDOMToken_fixed.set(this, void 0);
        this._astToken = astToken;
        tslib_1.__classPrivateFieldSet(this, _MLDOMToken_raw, astToken.raw, "f");
        tslib_1.__classPrivateFieldSet(this, _MLDOMToken_fixed, astToken.raw, "f");
        this.uuid = astToken.uuid;
        tslib_1.__classPrivateFieldSet(this, _MLDOMToken_startLine, astToken.startLine, "f");
        tslib_1.__classPrivateFieldSet(this, _MLDOMToken_endLine, astToken.endLine, "f");
        tslib_1.__classPrivateFieldSet(this, _MLDOMToken_startCol, astToken.startCol, "f");
        tslib_1.__classPrivateFieldSet(this, _MLDOMToken_endCol, astToken.endCol, "f");
        tslib_1.__classPrivateFieldSet(this, _MLDOMToken_startOffset, astToken.startOffset, "f");
        tslib_1.__classPrivateFieldSet(this, _MLDOMToken_endOffset, astToken.endOffset, "f");
    }
    get originRaw() {
        return tslib_1.__classPrivateFieldGet(this, _MLDOMToken_raw, "f");
    }
    get raw() {
        return tslib_1.__classPrivateFieldGet(this, _MLDOMToken_fixed, "f");
    }
    get startOffset() {
        return tslib_1.__classPrivateFieldGet(this, _MLDOMToken_startOffset, "f");
    }
    get endOffset() {
        return tslib_1.__classPrivateFieldGet(this, _MLDOMToken_endOffset, "f");
    }
    get startLine() {
        return tslib_1.__classPrivateFieldGet(this, _MLDOMToken_startLine, "f");
    }
    get endLine() {
        return tslib_1.__classPrivateFieldGet(this, _MLDOMToken_endLine, "f");
    }
    get startCol() {
        return tslib_1.__classPrivateFieldGet(this, _MLDOMToken_startCol, "f");
    }
    get endCol() {
        return tslib_1.__classPrivateFieldGet(this, _MLDOMToken_endCol, "f");
    }
    fix(raw) {
        tslib_1.__classPrivateFieldSet(this, _MLDOMToken_fixed, raw, "f");
    }
}
exports.default = MLDOMToken;
_MLDOMToken_startLine = new WeakMap(), _MLDOMToken_endLine = new WeakMap(), _MLDOMToken_startCol = new WeakMap(), _MLDOMToken_endCol = new WeakMap(), _MLDOMToken_startOffset = new WeakMap(), _MLDOMToken_endOffset = new WeakMap(), _MLDOMToken_raw = new WeakMap(), _MLDOMToken_fixed = new WeakMap();
