// eslint-plugin-markup/lib/rules/markup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.markup = void 0;
const synckit_1 = require("synckit");
const workerPath = require.resolve('../worker');
// call `creatSyncFn` lazily for performance, it is already cached inside, related #31
const _ = {
    get markuplintSync() {
        return (0, synckit_1.createSyncFn)(workerPath);
    },
};
exports.markup = {
    meta: {
        fixable: 'code',
        type: 'problem',
    },
    create(context) {
        const filename = context.getFilename();
        const sourceText = context.getSourceCode().text;
        const markuplintOptions = {
            sourceCode: sourceText,
            name: filename,
        };
        const runMarkuplint = (fix) => _.markuplintSync(markuplintOptions, fix);
        return {
            Program() {
                const { violations } = runMarkuplint();
                if (violations.length === 0) {
                    return;
                }
                let fixed = 0;
                for (const { ruleId, severity, message, line, col } of violations) {
                    context.report({
                        message: JSON.stringify({ severity, message, ruleId }),
                        loc: {
                            line,
                            // ! eslint ast column is 0-indexed, but markuplint is 1-indexed
                            column: col - 1,
                        },
                        fix() {
                            if (fixed++) {
                                return null;
                            }
                            const { fixedCode } = runMarkuplint(true);
                            return sourceText === fixedCode
                                ? null
                                : {
                                    range: [0, sourceText.length],
                                    text: fixedCode,
                                };
                        },
                    });
                }
            },
        };
    },
};
