// @markuplint/types/lib/match-result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unmatched = exports.matched = exports.matches = void 0;
function matches(checker, options) {
    return (value) => {
        const valid = checker(value);
        if (!valid && !value) {
            return unmatched(value, 'empty-token', options);
        }
        return valid ? matched() : unmatched(value, options === null || options === void 0 ? void 0 : options.reason, options);
    };
}
exports.matches = matches;
function matched() {
    return {
        matched: true,
    };
}
exports.matched = matched;
function unmatched(value, reason, options) {
    return {
        ...options,
        matched: false,
        ref: (options === null || options === void 0 ? void 0 : options.ref) || null,
        raw: value,
        offset: 0,
        length: value.length,
        line: 1,
        column: 1,
        reason: reason !== null && reason !== void 0 ? reason : 'syntax-error',
    };
}
exports.unmatched = unmatched;
