// @markuplint/html-parser/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.attrTokenizer = exports.parseRawTag = exports.getNamespace = exports.removeDeprecatedNode = exports.flattenNodes = exports.parse = exports.isDocumentFragment = void 0;
var is_document_fragment_1 = require("./is-document-fragment");
Object.defineProperty(exports, "isDocumentFragment", { enumerable: true, get: function () { return __importDefault(is_document_fragment_1).default; } });
var parse_1 = require("./parse");
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return parse_1.parse; } });
var flatten_nodes_1 = require("./flatten-nodes");
Object.defineProperty(exports, "flattenNodes", { enumerable: true, get: function () { return flatten_nodes_1.flattenNodes; } });
var remove_deprecated_node_1 = require("./remove-deprecated-node");
Object.defineProperty(exports, "removeDeprecatedNode", { enumerable: true, get: function () { return remove_deprecated_node_1.removeDeprecatedNode; } });
var get_namespace_1 = require("./get-namespace");
Object.defineProperty(exports, "getNamespace", { enumerable: true, get: function () { return get_namespace_1.getNamespace; } });
var parse_raw_tag_1 = require("./parse-raw-tag");
Object.defineProperty(exports, "parseRawTag", { enumerable: true, get: function () { return __importDefault(parse_raw_tag_1).default; } });
var attr_tokenizer_1 = require("./attr-tokenizer");
Object.defineProperty(exports, "attrTokenizer", { enumerable: true, get: function () { return __importDefault(attr_tokenizer_1).default; } });
